requirements/updater.sh

```diff
@@ -0,0 +1,76 @@
+#!/bin/sh
+set -ue
+
+export LC_ALL=C
+requirements_in="$(readlink -f ./requirements.in)"
+requirements_ansible_in="$(readlink -f ./requirements_ansible.in)"
+requirements="$(readlink -f ./requirements.txt)"
+requirements_ansible="$(readlink -f ./requirements_ansible.txt)"
+pip_compile="pip-compile --quiet -r --allow-unsafe"
+
+_cleanup() {
+  cd /
+  test "${KEEP_TMP:-0}" = 1 || rm -rf "${_tmp}"
+}
+
+install_deps() {
+  pip install pip --upgrade
+  pip install pip-tools
+}
+
+generate_requirements_v3() {
+  venv="./venv3"
+  python3 -m venv "${venv}"
+  # shellcheck disable=SC1090
+  . "${venv}/bin/activate"
+
+  install_deps
+
+  ${pip_compile} --output-file requirements.txt "${requirements_in}"
+  ${pip_compile} --output-file requirements_ansible_py3.txt "${requirements_ansible_in}"
+}
+
+generate_requirements_v2() {
+  venv="./venv2"
+  virtualenv -p python2 "${venv}"
+  # shellcheck disable=SC1090
+  PS1="" . "${venv}/bin/activate"
+
+  install_deps
+
+  ${pip_compile} --output-file requirements_ansible.txt "${requirements_ansible_in}"
+}
+
+generate_patch() {
+  a="requirements_ansible_py3.txt"
+  b="requirements_ansible.txt"
+
+  # most elegant/quick solution I could come up for now
+  out="$(diff --ignore-matching-lines='^#' --unified "${a}" "${b}" | \
+    awk '{ if (/^+\w/){ $2="; python_version < \"3\" #"; print;} else print; }' | \
+    sed 's/ ;/;/g')"
+  test -n "${out}"
+  echo "${out}"
+}
+
+main() {
+  _tmp="$(mktemp -d --suffix .awx-requirements XXXX -p /tmp)"
+
+  if [ "$1" = "upgrade" ]; then
+      pip_compile="${pip_compile} --upgrade"
```

Comment:
I'm perfectly happy with this. Related comment I left:

https://github.com/ansible/awx/pull/4123/files#r298685713

The only thing is that I _think_ within this PR you have so-far used `./upgrade.sh upgrade`, where as only `./upgrade.sh` would be better.

Reply:
Answering from my "private" GitHub account (thank you for your quick response):

No, merge request was run without upgrade. The dependencies in the requirements.txt got updated because of [ansible.in](https://github.com/ansible/awx/blob/d438a93fd2da796b546e2445f869d0b3ed1427ff/requirements/requirements_ansible.in) not having pinned certain dependencies.

I mostly splitted it up into two commits so I would get to see what actually changes on bumping the openstacksdk version:

- This happens if I just run `./upgrade.sh` without `upgrade` (or `--upgrade, -U` with `pip-compile`) => https://github.com/ansible/awx/pull/4123/commits/3db9f2546449e39f2c27394f74878fe4dd435417
- This happens when I just bump openstacksdk version afterwards https://github.com/ansible/awx/pull/4123/commits/d6e0bc85a12a1d462666fc2cab214bace31ffee6

In theory it would be enough to bump the versions as required by: [openstacksdk requirements.txt](https://github.com/openstack/openstacksdk/blob/442f6bbde3563875c5e19084cb28764736f4b6dc/requirements.txt) but this is also quite a lot of work.

Reply:
I've just tried it again - the dependencies are always getting updated no matter -U flag or not.

Reply:
I started trying to run your script here, and I think you're missing something from the expected use pattern, which could explain why `-U` doesn't seem to be respected.

I see that you create a new temporary directory and run the tool from there. This is fine, but when you do this, you are not giving it a reference `requirements.txt` (for instance) file. You are running

```
pip-compile --no-header --quiet -r --allow-unsafe --output-file requirements.txt /awx_devel/requirements/requirements.in
```

in a new folder where the file `requirements.txt` doesn't exist. How could it possible to _not_ upgrade? It doesn't _know_ what the old versions are. I think maybe if you were to copy over the old `requirements.txt` to the temporary directory before you do this, then it would be able to recognize existing versions and avoid upgrading those not related to the change being made.

Reply:
I haven't heard of pip-compile before, so excuse my confusion about how to exactly use the tool. 


```
$ pip-compile --help
  -o, --output-file FILENAME      Output file name. Required if more than one
                                  input file is given. Will be derived from
                                  input file otherwise.
```

Is a little bit confusing, I did not expect that it would pre-parse the output before writing into it.

I've just tried specifying an existing requirements.txt and it does indeed keep the pinned versions and only bump if I explicitly specify `--upgrade`.
 
Got to fix the script with my wrong assumptions...

Reply:
Thank you for the pointer, this change takes the copying of requirements into account https://github.com/ansible/awx/pull/4123/commits/0049ed6e5e82afda7da9a670914a4290cd40344b

As you can see, the versions aren't getting bumped anymore.
https://github.com/ansible/awx/pull/4123/commits/3eed82037954a24d6ac7bc85c4bdfe807598a2e2

plus bumping the openstacksdk version now as expected only causes a small footprint change
https://github.com/ansible/awx/pull/4123/commits/272f466082df82f318d0e59235895ac9fbdef075

How do you want to proceed?

I could just squash the commits

I could create two seperate merge requests (one with the helper script and the other one with the bumped openstacksdk)

Reply:
We would like commits squashed so that the churn doesn't affect the history or blame of the `*.txt` files. Other than that, you can keep them as 1 commit or multiple commits to logically separate the tool and the openstacksdk bump. I'd say that's up to you, but I'd suggest multi-line commit message if it's not a short message.

Reply:
Ok, I'll squash them into two or three commits...